# Exercise 4.4: Data Wrangling and Subsetting

## Import Libraries

In [1]:
import pandas as pd
import numpy as np
import os

## Loading the Data

In [2]:
# Define the path to the data files
path = '/Users/aaronkibler/CF Project 4 - Instacart Basket Analysis/02 Data'

In [3]:
df_ords = pd.read_csv(os.path.join(path, 'Original Data', 'orders.csv'))
df_prods = pd.read_csv(os.path.join(path, 'Original Data', 'products.csv'))
df_deps = pd.read_csv(os.path.join(path, 'Original Data','departments.csv'))

In [4]:
# Checking "orders.csv" data is loaded
print(df_ords.head())
print(df_ords.info())
print(df_ords.shape)

   order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0  
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                int64  
 1   user_id                 int64  
 2   eval_set                object 
 3   order_number            int64  
 4   order_dow               int64 

In [5]:
# Checking "products.csv" data is loaded
print(df_prods.head())
print(df_prods.info())
print(df_prods.shape)

   product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3  
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49693 entries, 0 to 49692
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   product_id     49693 non-null  int64  
 1   product_name   49677 non-null  object 
 2   aisle_id       49693 non-null  int64  
 3   department_id  49693 non-null  int64

In [6]:
# Checking "departments.csv" data is loaded
print(df_deps.head())
print(df_deps.info())
print(df_deps.shape)

  department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1 entries, 0 to 0
Data columns (total 22 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   department_id  1 non-null      object
 1   1              1 non-null      object
 2   2              1 non-null      object
 3   3              1 non-null      object
 4   4              1 non-null      object
 5   5              1 non-null      object
 6   6              1 non-null      object
 7   7              1 non-null      obj

## Dropping Unnecessary Columns

In [7]:
# Removes the "eval_set" column from the dataframe, but doesn't save the change
df_ords.drop(columns=['eval_set'])

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

In [8]:
# Removes the "eval_set" column and saves the change by overwriting df_ords
df_ords = df_ords.drop(columns=['eval_set'])

In [9]:
# Check that the column is dropped
print(df_ords.columns)

Index(['order_id', 'user_id', 'order_number', 'order_dow', 'order_hour_of_day',
       'days_since_prior_order'],
      dtype='object')


In [10]:
# Counting the occurrences of each unique value in the "days_since_prior_order" column
value_counts = df_ords['days_since_prior_order'].value_counts(dropna=False)

In [11]:
# Checking the frequency distribution of "days_since_prior_order"
print(value_counts)

days_since_prior_order
30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: count, dtype: int64


## Renaming Columns

In [12]:
# Rename "order_dow" to "orders_day_of_week"
df_ords.rename(columns={'order_dow': 'orders_day_of_week'}, inplace=True)

In [13]:
# Check the columns to verify the change
print(df_ords.columns)

Index(['order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_prior_order'],
      dtype='object')


## Changing Data Types

In [14]:
# Check the descriptive statistics of the columns to see which ones need to have their data type changed
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [15]:
# Change the data type of the identifier column "order_id" to string
df_ords['order_id'] = df_ords['order_id'].astype('str')

In [16]:
# Check the changes made to the data type
df_ords['order_id'].dtype

dtype('O')

## Transposing Data

In [17]:
# Import the data without the index column
df_deps = pd.read_csv(os.path.join(path, 'Original data', 'departments.csv'), index_col = False)

In [18]:
# Checking the first few rows of the "departments.csv" to understand its structure
print(df_deps.head)

<bound method NDFrame.head of   department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]>


In [19]:
# Transpose the "departments.csv" to convert it from wide to long format
df_dep_t = df_deps.T

In [20]:
# Checking the transposed dataframe
print(df_dep_t)

                             0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing


In [21]:
# Resetting the index of the transposed dataframe, which also turns the old index into a column
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [22]:
# Checking the dataframe after resetting the index to verify changes
print(df_dep_t)

                             0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing


In [23]:
# Extracting the first row to use as the new header (column names)
# "iloc[0]" selects the first row
new_header = df_dep_t.iloc[0]

In [24]:
# Removing the first row from the dataframe since it's now the header
# "[1:]" skips the first row
df_dep_t_new = df_dep_t[1:]

In [25]:
# Set the header row as the dataframe header
df_dep_t_new.columns = new_header

In [26]:
# Checking the cleaned dataframe to verify the new headers
print(df_dep_t_new)

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing


## Creating a Data Dictionary

In [27]:
# Convert the "df_dep_t_new" dataframe into a data dictionary (data_dict)
# "index" argument will use the numbered rows as the key values for the entries in the dictionary
data_dict = df_dep_t_new.to_dict('index')

In [28]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [29]:
# Check the "products.csv" data
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [30]:
# Use the data dictionary to determine the department_id of "19"
print(data_dict.get('19'))

{'department': 'snacks'}


## Subsetting Data

In [31]:
# Checking the "department_id" column of the "df_prods" dataframe for entries equal to "19"
df_prods['department_id']==19

0         True
1        False
2        False
3        False
4        False
         ...  
49688    False
49689    False
49690    False
49691    False
49692    False
Name: department_id, Length: 49693, dtype: bool

In [32]:
# "df_prods[df_prods['department_id'] == 19]" will create a new dataframe that shows the subset of "df_prods" whose "department_id" = 19
# We can assign this new dataframe the name "df_snacks"
df_snacks = df_prods[df_prods['department_id'] == 19]

In [33]:
# Check the new dataframe to verify the results
print(df_snacks.head())

    product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5  


In [34]:
# An alternative method using "loc" function to create a second dataframe (df_snacks_2) of the same subset
df_snacks_2 = df_prods.loc[df_prods['department_id'] == 19]

In [35]:
# Checking the new dataframe for "df_snacks_2"
df_snacks_2.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

In [36]:
# A third method, which is a variation of the "loc" function above:
df_snacks_3 = df_prods.loc[df_prods['department_id'].isin([19])]

In [37]:
# Checking the new dataframe for "df_snacks_3"
df_snacks_3.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

## Exporting DataFrames

In [38]:
# The following example code is an example of how to export the dataframes into the "Prepared Data" folder:
# df_ords.to_csv(os.path.join(path, 'Prepared Data', 'orders_wrangled.csv'))

# Tasks 4.4

## 2. Find another identifier variable in "df_ords" dataframe that can be changed to a more suitable format

In [39]:
# Check for additional columns in "df_ords" that need to have their data type changed
df_ords.describe()

user_id  order_number  orders_day_of_week  order_hour_of_day  \
count  3.421083e+06  3.421083e+06        3.421083e+06       3.421083e+06   
mean   1.029782e+05  1.715486e+01        2.776219e+00       1.345202e+01   
std    5.953372e+04  1.773316e+01        2.046829e+00       4.226088e+00   
min    1.000000e+00  1.000000e+00        0.000000e+00       0.000000e+00   
25%    5.139400e+04  5.000000e+00        1.000000e+00       1.000000e+01   
50%    1.026890e+05  1.100000e+01        3.000000e+00       1.300000e+01   
75%    1.543850e+05  2.300000e+01        5.000000e+00       1.600000e+01   
max    2.062090e+05  1.000000e+02        6.000000e+00       2.300000e+01   

       days_since_prior_order  
count            3.214874e+06  
mean             1.111484e+01  
std              9.206737e+00  
min              0.000000e+00  
25%              4.000000e+00  
50%              7.000000e+00  
75%              1.500000e+01  
max              3.000000e+01

In [40]:
# Change the data type of the identifier column "user_id" to string
df_ords['user_id'] = df_ords['user_id'].astype('str')

In [41]:
# Check the changes made to the data type
df_ords['user_id'].dtype

dtype('O')

## 3. Look for a variable in the "df_ords" dataframe with an unintuitive name and change its name without overwriting the dataframe.

In [42]:
# Check the names of the columns in "df_ords"
print(df_ords.columns)

Index(['order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_prior_order'],
      dtype='object')


In [43]:
# Change "days_since_prior_order" to "days_since_last_order" without overwriting the original
df_ords_new = df_ords.rename(columns={'days_since_prior_order': 'days_since_last_order'}, inplace=False)

In [44]:
# Check if the original column name remains unchanged
print("Original dataframe columns:", df_ords.columns)
print("New dataframe columns:", df_ords_new.columns)

Original dataframe columns: Index(['order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_prior_order'],
      dtype='object')
New dataframe columns: Index(['order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_last_order'],
      dtype='object')


## 4. What is the busiest hour for placing orders? Find the frequency of the corresponding variable and share your findings.

In [45]:
# From the previous task, we already know the column of interest is "order_hour_of_day" from "df_ords"
# The frequency can be found using the "value_counts" function
df_ords['order_hour_of_day'].value_counts(dropna = True)

order_hour_of_day
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64

### The busiest hour for placing orders is 10 am.

## 5. Determine the meaning behind a value of 4 in the "department_id" column within the df_prods dataframe using a data dictionary.

In [46]:
# Check number 4 department_id in the data dictionary
print(data_dict.get('4'))

{'department': 'produce'}


## 6. The sales team wants to know more about breakfast item sales. Create a subset containing only the required information.

In [47]:
# Check which department_id coresponds with breakfast items
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [48]:
# Create the new dataframe with the subset of breakfast items
df_breakfast = df_prods[df_prods['department_id']==14]

In [49]:
# Check the new dataframe to verify the results
df_breakfast.head()

product_id                                      product_name  aisle_id  \
27           28                                 Wheat Chex Cereal       121   
33           34                                               NaN       121   
67           68                           Pancake Mix, Buttermilk       130   
89           90                                      Smorz Cereal       121   
210         211  Gluten Free Organic Cereal Coconut Maple Vanilla       130   

     department_id  prices  
27              14    10.1  
33              14    12.2  
67              14    13.7  
89              14     3.9  
210             14     3.6

## 7. Find the details about products that customers might use to throw dinner parties. Your task is to find all observations from the entire dataframe that include items from the following departments: alcohol, deli, beverages, and meat/seafood. You’ll need to present this subset to your client.

In [50]:
# Check the data dictionary for the required departments
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [51]:
# The department_id for alcohol = 5, beverages = 7, meat/seafood = 12, and deli = 20
# Create a subset of the required departments
df_dinner_party = df_prods[df_prods['department_id'].isin([5, 7, 12, 20])]

In [52]:
# Check the new dataframe to verify the results
df_dinner_party.head()

product_id                                    product_name  aisle_id  \
2            3            Robust Golden Unsweetened Oolong Tea        94   
6            7                  Pure Coconut Water With Orange        98   
9           10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10          11                               Peach Mango Juice        31   
16          17                               Rendered Duck Fat        35   

    department_id  prices  
2               7     4.5  
6               7     4.4  
9               7     8.4  
10              7     2.8  
16             12    17.1

## 8. It’s important that you keep track of total counts in your dataframes. How many rows does the last dataframe you created have?

In [53]:
# Check the dimensions of the dinner party dataframe
df_dinner_party.shape

(7650, 5)

### The dinner party dataframe has 7650 rows.

## 9. Someone from the data engineers team in Instacart thinks they’ve spotted something strange about the customer with a "user_id" of “1.” Extract all the information you can about this user.

In [54]:
# Create a subset from df_ords to include all the data associated with user_id = 1
df_ords_user_1 = df_ords[df_ords['user_id']=='1']

In [55]:
# Extract all the information on user_id 1 from the new dataframe
df_ords_user_1

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329       1             1                   2                  8   
1   2398795       1             2                   3                  7   
2    473747       1             3                   3                 12   
3   2254736       1             4                   4                  7   
4    431534       1             5                   4                 15   
5   3367565       1             6                   2                  7   
6    550135       1             7                   1                  9   
7   3108588       1             8                   1                 14   
8   2295261       1             9                   1                 16   
9   2550362       1            10                   4                  8   
10  1187899       1            11                   4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

## 10. You also need to provide some details about this user’s behavior. What basic stats can you provide based on the information you have?

In [56]:
# Check the descriptive statistics of the df_ords_user_1 dataframe
df_ords_user_1.describe()

order_number  orders_day_of_week  order_hour_of_day  \
count     11.000000           11.000000          11.000000   
mean       6.000000            2.636364          10.090909   
std        3.316625            1.286291           3.477198   
min        1.000000            1.000000           7.000000   
25%        3.500000            1.500000           7.500000   
50%        6.000000            3.000000           8.000000   
75%        8.500000            4.000000          13.000000   
max       11.000000            4.000000          16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

## Export your df_ords dataframe as “orders_wrangled.csv” in your “Prepared Data” folder.

In [57]:
# Perform a final check of the dataframe before exporting
print(df_ords.head())
print(df_ords.info())
print(df_ords.shape)

  order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0  2539329       1             1                   2                  8   
1  2398795       1             2                   3                  7   
2   473747       1             3                   3                 12   
3  2254736       1             4                   4                  7   
4   431534       1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0  
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 6 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                object 
 1   user_id                 object 
 2   order_number            int64  
 3   orders_day_of_week      int64  
 4   order_hour_of_day       int64  
 5   days_

In [58]:
df_ords.to_csv(os.path.join(path, 'Prepared Data', 'orders_wrangled.csv'))

## Export the df_dep_t_new dataframe as “departments_wrangled.csv” in your “Prepared Data” folder so that you have a “.csv” file of your departments data in the correct format.

In [59]:
# Perform a final check of the dataframe before exporting
print(df_dep_t_new.head())
print(df_dep_t_new.info())
print(df_dep_t_new.shape)

department_id department
1                 frozen
2                  other
3                 bakery
4                produce
5                alcohol
<class 'pandas.core.frame.DataFrame'>
Index: 21 entries, 1 to 21
Data columns (total 1 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   department  21 non-null     object
dtypes: object(1)
memory usage: 336.0+ bytes
None
(21, 1)


In [60]:
df_dep_t_new.to_csv(os.path.join(path, 'Prepared Data', 'departments_wrangled.csv'))

# End of Exercise 4.4